delimited-core/src/main/scala/net/tixxit/delimited/parser/InputBuffer.scala

```diff
@@ -58,29 +60,58 @@ final class InputBuffer(input: Input) {
     }
     */
 
-    var p = pos
-    var i = 0
-    var res = -2
-    while (res == -2) {
-      if (i >= slen) {
-        res = i
-      }
-      else if (p >= clen) {
+    if (slen == 1) {
+      // this is a common case
+      if (pos >= clen) {
         // we exhausted this chunk
-        res = if (input.isLast) 0 else -1
+        if (input.isLast) 0 else -1
       }
-      else if (str.charAt(i) != chunk.charAt(p)) {
-        res = 0
+      else if (str.charAt(0) != chunk.charAt(pos)) {
+        0
+      }
+      else 1
+    }
+    else {
+      if (clen - pos >= slen) {
+        // the maximum i can be is slen,
+        // so if clen - pos >= slen, we can't exhaust
+        if (chunk.startsWith(str, pos)) slen
+        else 0
+      }
+      else {
+        var p = pos
+        var i = 0
+        // we know that clen - pos < slen
+        // so we cannot contain the entire string
+        // but we could have the beggining of str
+        // and then run out of input
+        // we can exhaust the input
+        while (true) {
```

Comment:
I tried to clean this up by writing `while (p < clen)` moving the `p >= clen` check here, but that seems to somehow hurt performance slightly. Maybe java likes the `while(true)` construct?

Anyway, I've reverted just to follow the benchmarks.